tensorboard/data/provider.py

```diff
@@ -567,6 +590,78 @@ class Hyperparameter:
     ] = None
 
 
+@dataclasses.dataclass(frozen=True)
+class HyperparameterValue:
+    """A hyperparameter value.
+
+    Attributes:
+      hyperparameter_name: A string identifier for the hyperparameters. It
+        corresponds to the hyperparameter_name field in the Hyperparameter
+        class.
+      domain_type: A HyperparameterDomainType describing how we represent the
+        set of known values in the `domain` attribute.
+      value: The value of the hyperparameter.
+
+        If domain_type is INTERVAL or DISCRETE_FLOAT, value is a float.
+        If domain_type is DISCRETE_STRING, value is a str.
+        If domain_type is DISCRETE_BOOL, value is a bool.
+        If domain_type is unknown (None), value is None.
+    """
+
+    hyperparameter_name: str
+    domain_type: Union[HyperparameterDomainType, None] = None
+    value: Union[float, str, bool, None] = None
+
+
+@dataclasses.dataclass(frozen=True)
+class HyperparameterSessionRun:
+    """A single run in a HyperparameterSessionGroup.
+
+    Attributes:
+      experiment_id: The id of the experiment to which the run belongs.
+      run: The name of the run.
+    """
+
+    experiment_id: str
+    run: str
+
+
+@dataclasses.dataclass(frozen=True)
+class HyperparameterSessionGroup:
+    """A group of runs logically executed together with the same hparam values.
+
+    The group of runs may have, for example, combined to generate and test a
+    single model or other artifacts.
+
+    We assume these groups of runs were executed with the same set of
+    hyperparameter values. However, having the same set of hyperparameter values
+    is not sufficient to be considered part of the same group -- different
+    groups can exist with the same hyperparameter values.
+
+    Attributes:
+      root: A descriptor of the common ancestor of all sessions in this
+        group.
+
+        In the case where the group contains all runs in the experiment, this
+        would just be a HyperparameterSessionRun with the experiment_id property
+        set to the experiment's id but run property set to empty.
+
+        In the case where the group contains a subset of runs in the experiment,
+        this would be a HyperparameterSessionRun with the experiment_id property
+        set and the run property set to the largest common prefix for runs.
```

Comment:
What if there's no common prefix, for example:
exp A has runs:
- pretrain
- train
- eval

where `pretrain` and `train` belong to one group and `eval` belongs to another, would `root` be an empty string in this case?

Reply:
I suggest:

The first group be HyperparameterSessionRun(experiment_id=expA_id, root='')

The second group by HyperparameterSessionRun(experiment_id=expA_id, root='eval')

I don't think it will be a problem if one root appears to be a subset of another root.